### PR TITLE
feature/VDX-178/pushed_auth_request

### PR DESCRIPTION
**Pull request by maikel-maas**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
   "license": "Apache-2.0",
   "private": true,
   "scripts": {
+    "start": "ts-node packages/issuer/lib/AuthServer.ts",
```
Comment (nklomp) [inline package.json:9]: Please move that to the actual module. The REST API for instance will move to another module. So make it the concern of these modules themselves

File: packages/client/lib/OpenID4VCIClient.ts [hunk 2]
```diff
@@ -136,7 +137,46 @@ export class OpenID4VCIClient {
...
 
-  private handleAuthorizationDetails(authorizationDetails?: AuthDetails | AuthDetails[]): AuthDetails | AuthDetails[] | undefined {
+  public async acquirePushedAuthorizationRequestURI({
+    clientId,
+    codeChallengeMethod,
+    codeChallenge,
+    authorizationDetails,
+    redirectUri,
+    scope,
+  }: AuthRequestOpts): Promise<OpenIDResponse<unknown>> {
```
Comment (nklomp) [inline packages/client/lib/OpenID4VCIClient.ts:147]: Why the unknown? You know what payload type to expect right?

File: packages/common/lib/types/Generic.types.ts [hunk 3]
```diff
@@ -202,5 +202,7 @@ export interface EndpointMetadata {
   token_endpoint: string;
   credential_endpoint: string;
   authorization_endpoint?: string;
+  pushed_authorization_request_endpoint?: string;
+  require_pushed_authorization_endpoint?: boolean;
```
Comment (nklomp) [inline packages/common/lib/types/Generic.types.ts:206]: I think I asked this already, but which spec does this value come from? This is a payload. We cannot makeup something ourselves here (if that is the case)

Didn't you happen to mean require_pushed_authorization_requests from https://www.rfc-editor.org/rfc/rfc9126#section-6

File: packages/common/lib/types/OpenID4VCIServerMetadata.ts [hunk 4]
```diff
@@ -12,6 +12,9 @@ export interface OpenID4VCIServerMetadata {
...
   // TODO: The above authorization_server being used in the wild, serves roughly the same purpose as the below spec compliant endpoint. Look at how to use authorization_server as authorization_endpoint in case it is present
   authorization_endpoint?: string;
+  pushed_authorization_request_endpoint?: string; // The URL of the pushed authorization request endpoint at which a client can post an authorization request to exchange for a request_uri value usable at the authorization server
+  // Note that the presence of pushed_authorization_request_endpoint is sufficient for a client to determine that it may use the PAR flow. A request_uri value obtained from the PAR endpoint is usable at the authorization endpoint regardless of other authorization server metadata such as request_uri_parameter_supported or require_request_uri_registration
+  require_pushed_authorization_endpoint?: boolean; // Boolean parameter indicating whether the authorization server accepts authorization request data only via PAR. If omitted, the default value is false.
```
Comment (nklomp) [inline packages/common/lib/types/OpenID4VCIServerMetadata.ts:17]: I think I asked this already, but which spec does this value come from? This is a payload. We cannot makeup something ourselves here (if that is the case)
Didn't you happen to mean require_pushed_authorization_requests from https://www.rfc-editor.org/rfc/rfc9126#section-5 ?

File: packages/issuer-rest/lib/api.ts [hunk 5]
```diff
@@ -97,37 +99,82 @@ export class RestAPI {
...
+        return res.status(400).json({
+          error: 'invalid_request',
+          error_description: message,
+        })
+      }
+      return next()
+    }
+
+    this.express.post('/par', handleHttpStatus400, (req: Request, res: Response) => {
+      // Fake client for testing, it needs to come from a registered client
```
Comment (nklomp) [inline packages/issuer-rest/lib/api.ts:145]: Please make this a fixme and create a ticket for it

File: packages/issuer-rest/lib/expressUtils.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { Request } from 'express'
+
+export const validateRequestBody = ({
```
Comment (nklomp) [inline packages/issuer-rest/lib/expressUtils.ts:3]: Either make the name more clear, or just throw an error from here (and catch it).

**Review by nklomp — CHANGES_REQUESTED** [review]
See some initial remarks. Need to do a more thorough review later

File: packages/common/lib/types/Generic.types.ts (hunk 3 above)
Comment (maikel-maas) [reply packages/common/lib/types/Generic.types.ts:206]: It's from [here](https://www.rfc-editor.org/rfc/rfc9126#name-authorization-server-metada) actually and it indicates whether the authorization server accepts authorization request data only via PAR.

File: packages/common/lib/types/OpenID4VCIServerMetadata.ts (hunk 4 above)
Comment (maikel-maas) [reply packages/common/lib/types/OpenID4VCIServerMetadata.ts:17]: Yes

File: package.json (hunk 1 above)
Comment (maikel-maas) [reply package.json]: Fixed!

File: packages/client/lib/OpenID4VCIClient.ts (hunk 2 above)
Comment (maikel-maas) [reply packages/client/lib/OpenID4VCIClient.ts]: Fixed!

File: packages/common/lib/types/Generic.types.ts (hunk 3 above)
Comment (maikel-maas) [reply packages/common/lib/types/Generic.types.ts:206]: It's from [section 5](https://www.rfc-editor.org/rfc/rfc9126#name-authorization-server-metada). It indicates whether the authorization server accepts authorization request data only via PAR

File: packages/issuer-rest/lib/api.ts (hunk 5 above)
Comment (maikel-maas) [reply packages/issuer-rest/lib/api.ts]: Fixed!

File: packages/issuer-rest/lib/expressUtils.ts (hunk 6 above)
Comment (maikel-maas) [reply packages/issuer-rest/lib/expressUtils.ts:3]: Fixed

File: packages/issuer-rest/lib/api.ts [hunk 7]
```diff
@@ -97,37 +99,83 @@ export class RestAPI {
...
+
+      // The scopes from the request need to be matched against the ones present in the registered client
+      if (!req.body.scope.split(',').every((scope: string) => client.scope.includes(scope))) {
+        return res.status(400).send({ error: 'invalid_scope', error_description: 'scope is not valid for the given client' })
+      }
+
+      //TODO Implement authorization_details verification
+
+      const uuid = uuidv4()
+      const requestUri = `urn:ietf:params:oauth:request_uri:${uuid}`
```
Comment (nklomp) [inline packages/issuer-rest/lib/api.ts:166]: The base URL for sure would also need to be optionally provided by the developer. Meaning there should be a configuration option fo the base URI (so everything in front of the uuid). This also needs a TODO, as in the future you even might want to bind the request uri to a specific client for instance

**Review by nklomp — CHANGES_REQUESTED** [review]
See remarks

File: packages/common/lib/types/Generic.types.ts (hunk 3 above)
Comment (nklomp) [reply packages/common/lib/types/Generic.types.ts:206]: Values should not be in this object to begin with. This is an object we create based on retrieval of several metadata endpoints. The value should be in the OAuth2 Server metadata, so also not in the IssuerMetadata. Will fix that

File: packages/common/lib/types/OpenID4VCIServerMetadata.ts (hunk 4 above)
Comment (nklomp) [reply packages/common/lib/types/OpenID4VCIServerMetadata.ts:17]: Values should not be in this object to begin with. This is an object we create based on retrieval of several metadata endpoints. The value should be in the OAuth2 Server metadata, so also not in the IssuerMetadata. Will fix that

**Review by nklomp — APPROVED** [review]

**Pull request merged**